### PR TITLE
Python march updates

### PR DESCRIPTION
**Pull request by kraigb**

### CODE REVIEW

**Comment by kraigb** [comment]
@gregvanl To merge when the next Python extension update goes out.
@brettcannon @DonJayamanne please review

Addresses #1529 #1527 #1500 plus a request for clarification from PR #1511

**Comment by brettcannon** [comment]
LGTM! I'll leave a comment and/or email when we make our release publicly available.

File: docs/python/debugging.md [hunk 1]
```diff
@@ -37,23 +37,27 @@ Standard configuration for `launch.json`:
...
+Identifies the type of debugger to use; leave this set to `python` for Python code.
 
-Points to the Python interpreter to be used for debugging purposes. The standard configuration uses the interpreter identified in the `python.pythonPath` setting by referring to `${config:python.pythonPath}`. To use a different interpreter, specify its path instead.
+### `request`
+
+Specifies the mode in which to start debugging:
+
+- `launch`: start the debugger on the file specified in `program`
+- `attach`: attach the debugger to an already running process. See [Remote debugging](#_remote-debugging) for an example.
+- `shell`: launches a separate program as specifies in `args`. See [Goodle App Engine debugging](#_google-app-engine-debugging) for an example.
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: Let me get back to you on this.

File: docs/python/debugging.md [hunk 2]
```diff
@@ -81,7 +103,7 @@ Specifies arguments to pass to the python program, for example:
 
 ### `stopOnEntry`
 
-When set to true, breaks the debugger at the first line of the program being debugged. Setting to false runs the program to the first breakpoint.
+When omitted or set to true, breaks the debugger at the first line of the program being debugged. Setting to false runs the program to the first breakpoint.
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: This is not true anymore. The value must be set to true. The current default value is `false`.

File: docs/python/debugging.md (hunk 1 above)
Comment (qubitron) [reply docs/python/debugging.md]: Typo: Goodle App Engine

File: docs/python/debugging.md [hunk 3]
```diff
@@ -95,7 +117,15 @@ Specifies how program output is displayed.
 
 ### `cwd`
 
-Specifies the current working directory for the program being debugged. The standard configuration uses `${workspaceFolder}` to use the project folder. If not specified, `cwd` defaults to the current folder open in VS Code.
+Specifies the current working directory for the debugger, which is the base folder for any relative paths used in code. If omitted, defaults to the folder of the file specified in `program`. A common setting is to use `${workspaceFolder}` (the folder open in VS Code) with or without additional subfolders.
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: Let me double check this logic and get back.

File: docs/python/debugging.md (hunk 1 above)
Comment (kraigb) [reply docs/python/debugging.md]: My bad...it should be under "type" as that's where it's used with the Google App Engine.

File: docs/python/debugging.md (hunk 3 above)
Comment (kraigb) [reply docs/python/debugging.md]: Please do...this seems to be the behavior from the tests I ran, but verifying from the code is better. :)

File: docs/python/debugging.md (hunk 2 above)
Comment (kraigb) [reply docs/python/debugging.md]: I'll fix.

File: docs/python/debugging.md (hunk 3 above)
Comment (DonJayamanne) [reply docs/python/debugging.md]: 1. If you open a workspace in VS Code, then the default is to use the `${workspaceFolder}`. 
2. If you do not open a workspace in VS Code, but attempt to debug an arbitrary python file then the default is the directory of the python file.

File: docs/python/debugging.md [hunk 4]
```diff
@@ -95,7 +117,15 @@ Specifies how program output is displayed.
...
+
+As an example, say `${workspaceFolder}` contains a `py_code` folder containing `app.py`, and a `data` folder containing `salaries.csv`. If you start the debugger on `py_code/app.py`, then the relative paths to the data file vary depending on the value of `cwd`:
+
+| cwd | Relative path to data file |
+| not specified (or `py_code`) | `../data/salaries.csv` |
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: Note: By default the value of `cwd` is `${worksapceFolder}`
Not correct, the current directory will be `${workspaceFolder}`.
If the file being debugged doesn't belong to the current workspace, then the value for `cwd` will default to the directory containing the file. Scenarios where a user opens an arbitrary python file and tries to debug it.

File: docs/python/debugging.md [hunk 5]
```diff
@@ -95,7 +117,15 @@ Specifies how program output is displayed.
 
 ### `cwd`
 
-Specifies the current working directory for the program being debugged. The standard configuration uses `${workspaceFolder}` to use the project folder. If not specified, `cwd` defaults to the current folder open in VS Code.
+Specifies the current working directory for the debugger, which is the base folder for any relative paths used in code. If omitted, defaults to the folder of the file specified in `program`. A common setting is to use `${workspaceFolder}` (the folder open in VS Code) with or without additional subfolders.
+
+As an example, say `${workspaceFolder}` contains a `py_code` folder containing `app.py`, and a `data` folder containing `salaries.csv`. If you start the debugger on `py_code/app.py`, then the relative paths to the data file vary depending on the value of `cwd`:
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: Note: Default value of `cwd` is always the root workspace folder, i.e. `${workspaceFolder}`.
If the user has specified something we use that, or if the file doesn't belong to a workspace, then we use the root directory of the file as the value for `cwd`.

File: docs/python/debugging.md [hunk 6]
```diff
@@ -95,7 +117,15 @@ Specifies how program output is displayed.
...
+
+| cwd | Relative path to data file |
+| not specified (or `py_code`) | `../data/salaries.csv` |
+| `${workspaceFolder}/py_code`) | `../data/salaries.csv` |
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: Yes

File: docs/python/debugging.md [hunk 7]
```diff
@@ -95,7 +117,15 @@ Specifies how program output is displayed.
...
+| `${workspaceFolder}` | `data/salaries.csv` |
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: Yes

File: docs/python/debugging.md [hunk 8]
```diff
@@ -95,7 +117,15 @@ Specifies how program output is displayed.
...
+| `${workspaceFolder}/data` | `salaries.csv` |
```
Comment (DonJayamanne) [inline docs/python/debugging.md]: Yes

Comment (DonJayamanne) [reply docs/python/debugging.md]: @kraigb 
Please feel free to download the latest version of the extension from here:
https://pvsc.blob.core.windows.net/extension-builds/ms-python-insiders.vsix

File: docs/python/debugging.md (hunk 3 above)
Comment (DonJayamanne) [reply docs/python/debugging.md]: @brettcannon I'm kinda having a brain freeze right now. Please could you confirm this.

Comment (brettcannon) [reply docs/python/debugging.md]: You can't debug an arbitrary file as VS Code demands a workspace, so I think you only need to care about option 1 and I think that's right. 😉

Comment (kraigb) [reply docs/python/debugging.md]: OK, I ran some more tests and confirmed ${workspaceFolder} as the default when cwd is omitted.

File: docs/python/debugging.md (hunk 5 above)
Comment (kraigb) [reply docs/python/debugging.md]: Take a look now.

**Comment by brettcannon** [comment]
@kraigb this can now be merged 🎉

**Comment by kraigb** [comment]
@gregvanl Ready to publish

**Comment by gregvanl** [comment]
@kraigb @brettcannon Published.

**Pull request merged**